src/Workspaces/Remote/ServiceHub/Host/RemoteAnalyzerAssemblyLoaderService.cs

```diff
@@ -16,16 +18,52 @@ namespace Microsoft.CodeAnalysis.Remote.Diagnostics
     [ExportWorkspaceService(typeof(IAnalyzerAssemblyLoaderProvider), WorkspaceKind.RemoteWorkspace), Shared]
     internal sealed class RemoteAnalyzerAssemblyLoaderService : IAnalyzerAssemblyLoaderProvider
     {
-        private readonly DefaultAnalyzerAssemblyLoader _loader = new DefaultAnalyzerAssemblyLoader();
-        private readonly ShadowCopyAnalyzerAssemblyLoader _shadowCopyLoader = new ShadowCopyAnalyzerAssemblyLoader(Path.Combine(Path.GetTempPath(), "VS", "AnalyzerAssemblyLoader"));
+        private readonly RemoteAnalyzerAssemblyLoader _loader;
+        private readonly ShadowCopyAnalyzerAssemblyLoader _shadowCopyLoader;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public RemoteAnalyzerAssemblyLoaderService()
         {
+            _loader = new(Path.GetDirectoryName(Path.GetFullPath(typeof(RemoteAnalyzerAssemblyLoader).GetTypeInfo().Assembly.Location)));
+            _shadowCopyLoader = new(Path.Combine(Path.GetTempPath(), "VS", "AnalyzerAssemblyLoader"));
         }
 
         public IAnalyzerAssemblyLoader GetLoader(in AnalyzerAssemblyLoaderOptions options)
             => options.ShadowCopy ? _shadowCopyLoader : _loader;
+
+        // For analyzers shipped in Roslyn, different set of assemblies might be used when running
+        // in-proc and OOP e.g. in-proc (VS) running on desktop clr and OOP running on ServiceHub .Net6
+        // host. We need to make sure to use the right one for OOP.
+        private class RemoteAnalyzerAssemblyLoader : DefaultAnalyzerAssemblyLoader
+        {
+            private readonly string? _baseDirectory;
+
+            protected override Assembly LoadImpl(string fullPath)
+                => base.LoadImpl(FixPath(fullPath));
+
+            public RemoteAnalyzerAssemblyLoader(string? baseDirectory)
```

Comment:
>? [](http://example.com/codeflow?start=54&length=1)

when is baseDirectory null?

Reply:
It shouldn't. I was mulling over whether to set it to null when not targeting .netcore (to avoid these file path operations)

Reply:
Fixed